ts-sdk/whirlpool/src/decreaseLiquidity.ts

```diff
@@ -0,0 +1,417 @@
+    lowerTickArrayAddress,
+    upperTickArrayAddress,
+  ]);
+
+  const lowerTick = lowerTickArray.data.ticks[getTickIndexInArray(position.data.tickLowerIndex, lowerTickArrayStartIndex, whirlpool.data.tickSpacing)];
+  const upperTick = upperTickArray.data.ticks[getTickIndexInArray(position.data.tickUpperIndex, upperTickArrayStartIndex, whirlpool.data.tickSpacing)];
+
+  const feesQuote = collectFeesQuote(
+    whirlpool.data,
+    position.data,
+    lowerTick,
+    upperTick,
+  );
+  const currentUnixTimestamp = BigInt(Math.floor(Date.now() / 1000));
+  const rewardsQuote = collectRewardsQuote(
+    whirlpool.data,
+    position.data,
+    lowerTick,
+    upperTick,
+    currentUnixTimestamp,
+  );
+
+  const { createInstructions, cleanupInstructions, tokenAccountAddresses } =
+    await prepareTokenAccountsInstructions(rpc, authority, [
+      whirlpool.data.tokenMintA,
+      whirlpool.data.tokenMintB,
+      whirlpool.data.rewardInfos[0].mint,
+      whirlpool.data.rewardInfos[1].mint,
+      whirlpool.data.rewardInfos[2].mint,
+    ]);
+
+  instructions.push(...createInstructions);
+
+  instructions.push(
+    getCollectFeesInstruction({
+      whirlpool: whirlpool.address,
+      positionAuthority: authority,
+      position: positionAddress[0],
+      positionTokenAccount,
+      tokenOwnerAccountA: tokenAccountAddresses[whirlpool.data.tokenMintA],
+      tokenOwnerAccountB: tokenAccountAddresses[whirlpool.data.tokenMintB],
+      tokenVaultA: whirlpool.data.tokenVaultA,
+      tokenVaultB: whirlpool.data.tokenVaultB,
+    }),
+  );
+
+  if (rewardsQuote.rewardOwed1 > 0) {
+    instructions.push(
+      getCollectRewardInstruction({
+        whirlpool: whirlpool.address,
+        positionAuthority: authority,
+        position: positionAddress[0],
+        positionTokenAccount,
+        rewardOwnerAccount:
+          tokenAccountAddresses[whirlpool.data.rewardInfos[0].mint],
+        rewardVault: whirlpool.data.rewardInfos[0].vault,
+        rewardIndex: 0,
+      }),
+    );
+  }
+
+  if (rewardsQuote.rewardOwed2 > 0) {
+    instructions.push(
+      getCollectRewardInstruction({
+        whirlpool: whirlpool.address,
+        positionAuthority: authority,
+        position: positionAddress[0],
+        positionTokenAccount,
+        rewardOwnerAccount:
+          tokenAccountAddresses[whirlpool.data.rewardInfos[1].mint],
+        rewardVault: whirlpool.data.rewardInfos[1].vault,
+        rewardIndex: 1,
+      }),
+    );
+  }
+
+  if (rewardsQuote.rewardOwed3 > 0) {
+    instructions.push(
+      getCollectRewardInstruction({
+        whirlpool: whirlpool.address,
+        positionAuthority: authority,
+        position: positionAddress[0],
+        positionTokenAccount,
+        rewardOwnerAccount:
+          tokenAccountAddresses[whirlpool.data.rewardInfos[2].mint],
+        rewardVault: whirlpool.data.rewardInfos[2].vault,
+        rewardIndex: 2,
+      }),
+    );
+  }
+
```

Comment:
If position.liquidity is zero, please skip decrease liquidity ix.
https://github.com/orca-so/whirlpools/blob/main/programs/whirlpool/src/instructions/decrease_liquidity.rs#L30